test/test_model_metadata.py

```diff
@@ -0,0 +1,274 @@
+                # Ignore table header
+                mock_header.return_value = ""
+                # Setup optional column return values
+                optional_column_outputs = [
+                    f"0.1" + " " * (INPUT_MIN_MAX_COLUMN_WIDTH - 3),
+                    f"2.0" + " " * (INPUT_MIN_MAX_COLUMN_WIDTH - 3),
+                    f"1.5" + " " * 16,
+                    " " * INPUT_MIN_MAX_COLUMN_WIDTH,
+                    " " * INPUT_MIN_MAX_COLUMN_WIDTH,
+                    "long_default_name" + " " * 2,
+                ]
+                mock_column.side_effect = optional_column_outputs
+                # Expected table
+                expected_table = (
+                    f"R Number"
+                    + " " * 2
+                    + f"number"
+                    + " " * (INPUT_TYPE_COLUMN_WIDTH - 6)
+                    + optional_column_outputs[0]
+                    + optional_column_outputs[1]
+                    + optional_column_outputs[2]
+                    + f"The reproduction number\n"
+                    + f"Setting"
+                    + " " * 3
+                    + f"string"
+                    + " " * (INPUT_TYPE_COLUMN_WIDTH - 6)
+                    + optional_column_outputs[3]
+                    + optional_column_outputs[4]
+                    + optional_column_outputs[5]
+                    + "Mode to run the model in\n"
+                )
+                # CALL
+                table_string = metadata.format_parameters()
+
+                # ASSERT
+                assert expected_table == table_string
+
+        class TestFormatDataslots:
+            """Test class to test the ModelMetadata.format_dataslots() functionality"""
+
+            def test_dataslots_is_formatted_properly(self, get_model_metadata_fixture):
+                # SETUP
+                metadata_dict = get_model_metadata_fixture
+                metadata = model_metadata.ModelMetadata(metadata_dict)
+                # Expected output
+                expected = (
+                    f"Name: Inputs\n"
+                    + f"Path in container: inputs/\n"
+                    + f"Required: True\n"
+                    + f"Default Datasets: \n"
+                    + f"Name: "
+                    + f"ID: 11111a1a-a111-11aa-a111-11aa11111aaa"
+                    + TAB_SPACE
+                    + f"Version ID: 21111a1a-a111-11aa-a111-11aa11111aaa"
+                    + TAB_SPACE
+                    + "\n"
+                )
+
+                # CALL
+                dataslot_string = metadata.format_dataslots()
+
+                # ASSERT
+                assert dataslot_string == expected
+
+        @patch("dafni_cli.model_metadata.outputs_table_header")
+        @patch("dafni_cli.model_metadata.optional_column")
+        class TestFormatOutputs:
+            """Test class to test the ModelMetadata.format_outputs() functionality"""
+
+            def test_outputs_table_is_formatted_properly(
```

Comment:
case where there are no datasets?

Reply:
Hopefully this won't be called when there aren't any outputs. And nothing changes when there aren't any datasets. Unless I've misunderstood...